masonry/src/widget/textbox.rs

```diff
@@ -27,278 +31,644 @@ const TEXTBOX_PADDING: f64 = 3.0;
+                        } else {
+                            txn.move_right();
+                        }
+                    }),
+                    Key::Named(NamedKey::ArrowUp) => self.editor.transact(fctx, lctx, |txn| {
+                        if shift {
+                            txn.select_up();
+                        } else {
+                            txn.move_up();
+                        }
+                    }),
+                    Key::Named(NamedKey::ArrowDown) => self.editor.transact(fctx, lctx, |txn| {
+                        if shift {
+                            txn.select_down();
+                        } else {
+                            txn.move_down();
+                        }
+                    }),
+                    Key::Named(NamedKey::Home) => self.editor.transact(fctx, lctx, |txn| {
+                        if action_mod {
+                            if shift {
+                                txn.select_to_text_start();
+                            } else {
+                                txn.move_to_text_start();
+                            }
+                        } else if shift {
+                            txn.select_to_line_start();
+                        } else {
+                            txn.move_to_line_start();
+                        }
+                    }),
+                    Key::Named(NamedKey::End) => self.editor.transact(fctx, lctx, |txn| {
+                        if action_mod {
+                            if shift {
+                                txn.select_to_text_end();
+                            } else {
+                                txn.move_to_text_end();
+                            }
+                        } else if shift {
+                            txn.select_to_line_end();
+                        } else {
+                            txn.move_to_line_end();
+                        }
+                    }),
+                    Key::Named(NamedKey::Delete) => self.editor.transact(fctx, lctx, |txn| {
+                        if action_mod {
+                            txn.delete_word();
+                        } else {
+                            txn.delete();
+                        }
+                    }),
+                    Key::Named(NamedKey::Backspace) => self.editor.transact(fctx, lctx, |txn| {
+                        if action_mod {
+                            txn.backdelete_word();
+                        } else {
+                            txn.backdelete();
+                        }
+                    }),
+                    Key::Named(NamedKey::Enter) => {
+                        ctx.submit_action(crate::Action::TextEntered(self.text().to_string()));
+                        return;
+                        // let (fctx, lctx) = ctx.text_contexts();
+                        // self.editor
+                        //     .transact(fctx, lctx, |txn| txn.insert_or_replace_selection("\n"));
+                    }
```

Comment:
At some point in the future we need to allow configuration of enter behavior. There are definitely instances where enter should submit. Others where it should add a new line. And in some applications, shift-enter should be new line.